packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -40,6 +40,59 @@ import { TokenFetchOptions } from "./tokenFetch";
 const afdUrlConnectExpirationMs = 6 * 60 * 60 * 1000; // 6 hours
 const lastAfdConnectionTimeMsKey = "LastAfdConnectionTimeMs";
 
+const localStorageAvailable = isLocalStorageAvailable();
+
+/**
+ * Helper to check the timestamp in localStorage (if available) indicating whether the cache is still valid.
+ */
+function isAfdCacheValid(): boolean {
+    if (localStorageAvailable) {
+        const lastAfdConnection = localStorage.getItem(lastAfdConnectionTimeMsKey);
+        if (lastAfdConnection !== null) {
+            const lastAfdTimeMs = Number(lastAfdConnection);
+            // If we have used the AFD URL within a certain amount of time in the past,
+            // then we should use it again.
+            if (!isNaN(lastAfdTimeMs) && lastAfdTimeMs > 0
+                && Date.now() - lastAfdTimeMs <= afdUrlConnectExpirationMs) {
+                return true;
+            } else {
+                localStorage.removeItem(lastAfdConnectionTimeMsKey);
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+ * Test if we deal with NetworkErrorBasic object and if it has enough information to make a call
+ * If in doubt, allow retries
+ *
+ * @param error - error object
+ */
+function canRetryOnError(error: any) {
```

Comment:
similar API exists in delta manager. Can we move it somewhere such that all layers use same logic (and it is exported, such that consumers of our repo can use too?)

Reply:
Moved to driver-utils in latest